### PR TITLE
Sliding Sync FFI and DevEx fixes

### PR DESCRIPTION
**Pull request by gnunicorn**
This PR fixes a few aspects ontop of the sliding sync FFI:

- [x] enable to-device messages via event-handler to decouple it from sync v2 / make it available to sliding sync messages
- [x] process sliding-sync errors same way as v2 in ffi crate
- [x] continue-after error for sliding-sync requests to continue the loop if even an error was found but polled again
- [x] parallelize olm-outgoing requests and sliding sync polling (waiting for it meant delaying time to first-view) //cc @poljar

### CODE REVIEW

File: crates/matrix-sdk/src/error.rs [hunk 1]
```diff
@@ -193,6 +193,11 @@ pub enum Error {
     #[error(transparent)]
     ImageError(#[from] ImageError),
 
+    /// An error occurred during decryption.
```
Comment (jplatte) [inline crates/matrix-sdk/src/error.rs:196]: This seems wrong.

File: crates/matrix-sdk/src/sliding_sync.rs [hunk 2]
```diff
@@ -878,17 +910,24 @@ impl SlidingSyncView {
...
+                        .ok_or_else(|| {
+                            Error::BadResponse(
+                                "`range` must be present for Sync and Update operation".to_owned(),
+                            )
+                        })?
                         .0
-                        .try_into()?;
+                        .try_into()
+                        .map_err(|e| {
+                            Error::BadResponse(format!("`range` not a valid int: {:}", e))
```
Comment (jplatte) [inline crates/matrix-sdk/src/sliding_sync.rs:929]: In addition to what clippy says, the `:` here is also a no-op AFAICT. So I think this should be

```suggestion
                            Error::BadResponse(format!("`range` not a valid int: {e}"))
```

File: crates/matrix-sdk/src/sliding_sync.rs [hunk 3]
```diff
@@ -550,14 +557,39 @@ impl SlidingSync {
...
-                    .await?;
+
+                let req = client.send_with_homeserver(req, None, self.homeserver.as_ref().map(ToString::to_string));
+
+                #[cfg(feature = "e2e-encryption")]
+                let resp_res = {
+                    let (e2ee_uploads, resp) = futures_util::join!(client.send_outgoing_requests(), req);
+                    if let Err(e) = e2ee_uploads {
+                        tracing::error!(error = ?e, "Error while sending outgoing E2EE requests");
+                    };
```
Comment (jplatte) [inline crates/matrix-sdk/src/sliding_sync.rs:568]:
```suggestion
                    }
```

File: crates/matrix-sdk/src/sliding_sync.rs [hunk 4]
```diff
@@ -550,14 +557,39 @@ impl SlidingSync {
...
+                    resp
+                };
+                #[cfg(not(feature = "e2e-encryption"))]
+                let resp_res = req.await;
+
+                let resp = match resp_res {
+                    Ok(r) => r,
+                    Err(e) => {
+                        yield Err(e.into());
+                        continue
```
Comment (jplatte) [inline crates/matrix-sdk/src/sliding_sync.rs:578]:
```suggestion
                        continue;
```

File: crates/matrix-sdk/src/sliding_sync.rs [hunk 5]
```diff
@@ -550,14 +557,39 @@ impl SlidingSync {
...
+                };
+
                 tracing::debug!("received");
 
-                let updates = self.handle_response(resp, &remaining_views).await?;
+                let updates =  match self.handle_response(resp, &remaining_views).await {
+                    Ok(r) => r,
+                    Err(e) => {
+                        yield Err(e.into());
+                        continue
```
Comment (jplatte) [inline crates/matrix-sdk/src/sliding_sync.rs:588]:
```suggestion
                        continue;
```

**Review by jplatte — APPROVED** [review]
Got some style nitpicks, but otherwise LGTM.

**Review by stefanceriu — APPROVED** [review]
Looks good to me too and I can confirm it's working properly in ElX

**Pull request merged**